### PR TITLE
Achievement details

### LINKED ISSUE #85: Collectibles grabber
Not an issue, but a question:
I need working on a proper tracker of achievements, minions, mounts, triple triad card and orchestrion rolls for my [tracker](https://simbiat.ru/fftracker) and I need to be able to grab up-to-date list of those. I mean, not those that a character has, but all of them. For achievements we can use existing functionality, but then they are not split into subcategories and Legacy ones are not shown for a lot of characters, so it's not reliable. So, I guess we need to use Eorzea Database for this, same as for other collectibles, although then, they will not have icons, but since I need their "howto", I guess I will need to load their pages as well.
So my question is whether I can use the parser somehow to parse the pages to get "items" URLs and then parse those URLS to get details? If not, do we need it built in here?

### LINKED ISSUE #91: Feasts number getting NULLed
New PvPTeam feature grabs "feasts" (number of Feast battles). For some reason, doing toArray on the result turns string '0' into NULL, but I do not see any kind of conversion like this in AbstractEntity. Similar thing is actually happening with "rank" and "rankicon" too, when character is a regular member. Also, "feasts" are shown as first element for the PvPTeamMembers object for some reason.

### PR DESCRIPTION
**Pull request by Simbiat**
Achievement details (to close #85 ), fix for `feasts` getting NULLed, fix for `feasts` order (to close #91 ), addition of URL to HTTP errors

### CODE REVIEW

**Comment by viion** [comment]
Random question, do you use `toArray()` ?

I'm considering adjusting some stuff with the api So instead of returning `grandcompany` it would be `GrandCompany`, title casing everything.

But this would break bc, so i might just have it as an option, eg `toArray(true)`

**Comment by Simbiat** [comment]
I do use it at the moment... but I "normalize" output either way, so maybe I do not even need to... I think the main reason I am using it is since I need to merge all company/linkshell/achievement pages into 1 and then iterate over them. Either way, even if you rename something - no biggie for me. I'll be testing either way and will adjust normalizer accordingly.

**Comment by viion** [comment]
Done some changes so you can pass `toArray(true)` and get capitalised indexes. Fixed a bunch of PSR issues and renamed some variables.

Can you make sure everything still OK your side and I will merge

**Comment by Simbiat** [comment]
Seems to be working fine for me after very small update of  Normalization function on my side.

**Comment by Simbiat** [comment]
1 concern though: for PvP teams, those are NOT servers, but actual Data Centers, so not sure if it's correct to name the variable "server". PvP teams are cross-server.

**Comment by viion** [comment]
Yes you are right, they can all be changed to `dataCenter`, i thought it was server because they were `getServer` and `setServer`

File: src/Lodestone/Api.php [hunk 1]
```diff
@@ -42,6 +42,9 @@
  */
 class Api
 {
+    private $useragent = '';
+    private $language = 'na';
+
```
Comment (viion) [inline src/Lodestone/Api.php]: These shouldn't be here? If they're relating to the query they should be in `HttpRequest`

File: src/Lodestone/Api.php [hunk 2]
```diff
@@ -75,8 +78,8 @@ public function searchCharacter($name, $server = false, $page = 1)
         $urlBuilder->add('worldname', $server);
         $urlBuilder->add('page', $page);
 
-        $url = Routes::LODESTONE_CHARACTERS_SEARCH_URL;
-        return (new CharacterSearch())->url($url . $urlBuilder->get())->parse();
+        $url = (new Routes($this->language))::$LODESTONE_CHARACTERS_SEARCH_URL;
+        return (new CharacterSearch())->url($url . $urlBuilder->get(), $this->useragent)->parse();
```
Comment (viion) [inline src/Lodestone/Api.php]: Why change this to a class and have to pass it along? Static classes means you only need to do it once. Make a `Routes::setLanguage` or `Routes::setLocale` which can be modified via the `Api` class before any queries.

File: src/Lodestone/Modules/Http/Routes.php [hunk 3]
```diff
@@ -14,53 +14,109 @@
...
-    const LODESTONE_ACHIEVEMENTS_CAT_URL = self::LODESTONE_URL . 'lodestone/character/%s/achievement/category/%s/';
-    const LODESTONE_ACHIEVEMENTS_DET_URL = self::LODESTONE_URL . 'lodestone/character/%s/achievement/detail/%s/';
+    static $LODESTONE_URL = '';
+    static $LODESTONE_CHARACTERS_URL = '';
+    static $LODESTONE_CHARACTERS_FRIENDS_URL = '';
+    static $LODESTONE_CHARACTERS_FOLLOWING_URL = '';
+    static $LODESTONE_CHARACTERS_SEARCH_URL = '';
+    static $LODESTONE_ACHIEVEMENTS_URL = '';
+    static $LODESTONE_ACHIEVEMENTS_CAT_URL = '';
+    static $LODESTONE_ACHIEVEMENTS_DET_URL = '';
```
Comment (viion) [inline src/Lodestone/Modules/Http/Routes.php]: ??? Why

File: src/Lodestone/Modules/Http/Routes.php [hunk 4]
```diff
@@ -14,53 +14,109 @@
...
+        // deep dungeon
+        self::$LODESTONE_DEEP_DUNGEON = self::$LODESTONE_URL .'lodestone/ranking/deepdungeon/';
+    
+        // world status
+        self::$LODESTONE_WORLD_STATUS = self::$LODESTONE_URL .'lodestone/worldstatus/';
+    
+        // devblog
+        self::$LODESTONE_DEV_BLOG = self::$LODESTONE_URL .'/pr/blog/atom.xml';
+    }
 }
```
Comment (viion) [inline src/Lodestone/Modules/Http/Routes.php]: Non of this is needed if you build your url during request

**Comment by viion** [comment]
I think i may merge this onto a branch for `Experimental` since there are a lot of changes going on with the logic and some bad habits are creeping up. The code is becoming very messy and can be simplified a massive amount.

**Comment by Simbiat** [comment]
But they are static. I am not sure they can be called in any other way, though, if we need them to be actually dynamic. It's a dirty implementation at the moment.

File: src/Lodestone/Modules/Http/Routes.php (hunk 4 above)
Comment (Simbiat) [reply src/Lodestone/Modules/Http/Routes.php]: do you you mean that I use language when calling url() functions only and simply drop it from Routes?

**Comment by viion** [comment]
I'm going to merge this into `experiment` which you can switch to in composer to get your changes. We can use that branch for pre-releases. I want to look into really simplifying a lot of the API or I may just make my own custom one as I need something much easier to maintain and faster to parse.

**Comment by Simbiat** [comment]
mmm.. Well, I was thinking of finding a way to make it a bit more transparent, too with those settings and stuff. I can suggest something like that:
Api.php has the settings (language and useragent), which can be changed after object creation. Those functions like getSomething do not actually get something, but, instead setup `$url` and `$type` in Api object (null by default). Then you call parse on Api object, which utilizes settings and calls actual HTML parser respective to the `$type` and updates `$result` in Api object, which is the object returned by parser.
Or something like that. While I was trying to add useragent and language, my main issue was that those objects had no idea about the parent, where they were called from, so I had to "drag" the parameters all the way down. If we use getSomething functions in Api *only* as a way to parametrize what we are looking for, it may make it a bit more transparent, I think. I will try to do that.

**Comment by viion** [comment]
Make them static and it won’t need to be dragged, it can be set and accessed at any point, having this duplicated all over is really bad and hard to maintain, its its configuration you can use static logic such as a static array in a class to hold the current state value.

**Pull request merged**